jupyterlab/jupyter-ai
# Add `metadata` field to agent messages

## Description

Adds a `metadata` field to the agent message model, which should stores message metadata returned by the LLM provider. In LangChain, all providers have the option to return a `generation_info` dictionary when yielding one or more [`Generation` objects](https://api.python.langchain.com/en/latest/outputs/langchain_core.outputs.generation.Generation.html#langchain_core.outputs.generation.Generation) while replying to a prompt. This PR makes that dictionary accessible via the `metadata` attribute in both the frontend & backend models for agent messages.

This dictionary may contain info about why an LLM finished generating a response (`stop_reason`) or append provider-specific data like a `session_id`. Some advanced custom use-cases like [implementing thumbs up/down feedback in a custom message footer](https://github.com/jupyterlab/jupyter-ai/pull/942) require this data to include in telemetry events.

In this implementation, this field is only set by `DefaultChatHandler`. For the time being, custom chat handlers have to mimic the implementation in this PR to include metadata in the returned message. In the future, the `metadata` field will be set automatically by `BaseChatHandler` so other chat handlers do not need to duplicate the implementation in their code. However, this requires refactoring `DefaultChatHandler` to migrate the streaming logic to `BaseChatHandler`.

## Demo

https://github.com/user-attachments/assets/5c70213d-89c6-4a94-9549-d05a82007131

## Testing

To replicate the above demo:

1. Open `packages/jupyter-ai/src/components/chat-messages.tsx`.
2. Add the below code to the top of `ChatMessageHeader`'s definition to log the message metadata object:

```tsx
export function ChatMessageHeader(props: ChatMessageHeaderProps): JSX.Element {
  const collaborators = useCollaboratorsContext();

  // new code here: vvv
  if (props.message.type === 'agent-stream' && props.message.complete) {
    console.log(props.message.metadata);
  }
  
  ...
}
```
3. Run `jlpm build` and restart JupyterLab.
4. Verify that message metadata is logged in the console after receiving a message. You can use `test-provider-with-streaming:test` model provided by `jupyter_ai_test` (installable via `jlpm dev-uninstall && jlpm dev-install` if not already installed). It should always set the message metadata to `{"test_metadata_field":"foobar"}`.
